### PR TITLE
Highlighter with Linker

### PR DESCRIPTION
**Pull request by HoshinoTented**
#464

### CODE REVIEW

**Comment by ice1000** [comment]
Oh my Goodness 🥵🥵🥵🥵🥵🥵🥵🥵🥵🥵🥵🥵♾️♾️♾️♾️♾️♾️♾️♾️♾️♾️♾️🎊🎊🎊🎊🎊🎊🎊🎊

**Comment by imkiva** [comment]
Oh my Goodness 🥵🥵🥵🥵🥵🥵🥵🥵🥵🥵🥵🥵♾️♾️♾️♾️♾️♾️♾️♾️♾️♾️♾️🎊🎊🎊🎊🎊🎊🎊🎊

File: cli/src/main/java/org/aya/cli/literate/Highlighter.java [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+      case Expr.RawProj rawProj -> {}
+      // only highlight keyword (but this will done in another place)
+      case Expr.Coe coe -> {}
+      case Expr.Do aDo -> {}
+      case Expr.Pi pi -> {}
+      case Expr.New aNew -> {}
+      case Expr.Match match -> {}
+      // no-op
+      case Expr.Array array -> {}
+      case Expr.Idiom idiom -> {}
```
Comment (imkiva) [inline cli/src/main/java/org/aya/cli/literate/Highlighter.java:147]: Maybe a `case Expr misc -> {}` ?

File: cli/src/main/java/org/aya/cli/literate/Highlighter.java [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+    switch (pattern) {
+      case Pattern.Bind bind -> highlightVarDef(bind.bind(), HighlightInfoType.DefKind.Local);
+      case Pattern.Ctor ctor -> {
+        var resolved = ctor.resolved();
+        highlightVarRef(resolved.sourcePos(), resolved.data(), HighlightInfoType.DefKind.Con);
+      }
+      case Pattern.Number number -> addInfo(number.sourcePos(), HighlightInfoType.LitInt.INSTANCE);
+      // no-op
+      case Pattern.BinOpSeq binOpSeq -> {}
+      case Pattern.CalmFace calmFace -> {}
```
Comment (imkiva) [inline cli/src/main/java/org/aya/cli/literate/Highlighter.java:178]: ditto?

File: cli/src/test/java/org/aya/cli/HighlighterTest.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+      | S n, m => S (add n m)
+      """;
+
+    highlightAndTest(code,
+      keyword(0, 4, "open"),
+      whatever(),
+      def(10, 13, "Nat", "nat", HighlightInfoType.DefKind.Data),
+      def(16, 17, "O", "zero", HighlightInfoType.DefKind.Con),
+      def(20, 21, "S", "suc", HighlightInfoType.DefKind.Con),
+      ref(22, 25, "Nat", "nat", HighlightInfoType.DefKind.Data),
```
Comment (imkiva) [inline cli/src/test/java/org/aya/cli/HighlighterTest.java:34]: This is really cool!!!!

File: cli/src/main/java/org/aya/cli/literate/Highlighter.java (hunk 1 above)
Comment (HoshinoTented) [reply cli/src/main/java/org/aya/cli/literate/Highlighter.java:147]: I hope that the compiler will remind you to impl the new case if you create a new subclass of Expr.

**Comment by HoshinoTented** [comment]
Oh my Goodness 🥵🥵🥵🥵🥵🥵🥵🥵🥵🥵🥵🥵♾️♾️♾️♾️♾️♾️♾️♾️♾️♾️♾️🎊🎊🎊🎊🎊🎊🎊🎊

File: cli/src/test/java/org/aya/cli/HighlighterTester.java [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+    @Override
+    public boolean hasNext() {
+      return !source.isEmpty();
+    }
+
+    @Override
+    public T next() {
+      return source.remove();
+    }
+  }
```
Comment (ice1000) [inline cli/src/test/java/org/aya/cli/HighlighterTester.java:57]: [`Seq.generateUntilNull`](https://github.com/Glavo/kala-common/commit/1f6315edb18613c576344c6346d5df3f1f291570)

File: cli/src/test/java/org/aya/cli/PlctReportTest.java [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+
+import org.aya.cli.plct.PLCTReport;
+import org.aya.cli.utils.MainArgs;
+import org.junit.jupiter.api.Test;
+
+public class PlctReportTest {
+  @Test public void generateReport() throws Exception {
+    var args = new MainArgs.PlctAction();
+    args.plctReport = true;
+    System.out.println(new PLCTReport().run(args));
```
Comment (imkiva) [inline cli/src/test/java/org/aya/cli/PlctReportTest.java:13]: This test takes 7s to finish in my machine. And the whole tests take 11s.

Comment (ice1000) [reply cli/src/test/java/org/aya/cli/PlctReportTest.java:13]: LMFAO

Comment (ice1000) [reply cli/src/test/java/org/aya/cli/PlctReportTest.java:13]: I guess we should enabled it only when `"true".equals(System.getenv("CI"))`

**Pull request merged**